### PR TITLE
Fix style mixing, for merging map properties on an empty target node

### LINKED ISSUE #618: Regression on style mixing uniform shader block
While adding uniform arrays, I discovered that we might have done a regression on uniform shader mixing (probably define too).

This currently fails on master, it seems that `heightglowline` doesn't get mixed properly for the block uniforms:

```yaml
     heightglow:
         base: polygons
         shaders:
             uniforms:
                  u_test: 800.0
             blocks:
                  color: "color.rgb += vec3(position.z / u_test);"
     heightglowline:
         base: lines
         mix: heightglow
```

Compilation error on the shader while compiling `heightglowline`:
```
ERROR shaderProgram.cpp:212: Compiling shader:
ERROR: 0:451: Use of undeclared identifier 'u_test'
```

We should consolidate the test case coverage for those.

### PR DESCRIPTION
**Pull request by tallytalwar**
Fixes #618

### CODE REVIEW

**Comment by karimnaaji** [comment]
I think it would make sense to have a test case that covers this

**Comment by tallytalwar** [comment]
Yup will add one.

**Comment by blair1618** [comment]
Yeahhhhh that function shouldn't return immediately if the map is null. We *do* still want to check if it is not a map, otherwise we'll crash below when we try to access elements by key.

**Comment by tallytalwar** [comment]
Test added! And this test fails on master branch.

**Comment by tallytalwar** [comment]
@blair1618, refer to the test case I added, we would like to merge (override the non-map target) with the mappings from the sources. I also do not see where will it crash?

File: core/src/scene/styleMixer.cpp [hunk 1]
```diff
@@ -201,9 +201,6 @@ void StyleMixer::mergeFieldTakingLast(std::string key, Node target, const std::v
 void StyleMixer::mergeMapFieldTakingLast(std::string key, Node target, const std::vector<Node>& sources) {
```
Comment (karimnaaji) [inline core/src/scene/styleMixer.cpp]: Just seen we were copying the keys here, we should use `const std::string&` instead. Can we update that at the same time?

**Comment by blair1618** [comment]
Yeah when `map` doesn't exist we want to override it with the values from `sources`, but if `map` exists and is a scalar or a list, then `map[subkey]` will crash.

**Comment by tallytalwar** [comment]
Yup, I updated my comment :D.

**Comment by blair1618** [comment]
So the change should just be `if (!map || !map.IsMap()) {` -> `if (map && !map.IsMap()) {`

File: core/src/scene/styleMixer.cpp (hunk 1 above)
Comment (blair1618) [inline core/src/scene/styleMixer.cpp]: It's probably move-constructed anyway, but couldn't hurt

Comment (karimnaaji) [inline core/src/scene/styleMixer.cpp]: I'm not convinced we should generally rely that much on that,
quick bench shows it's much faster to pass by const ref, even with optimization on:
https://gist.github.com/karimnaaji/d58afb771de5edfc98a1

```
~/Desktop/cpp-benchs  master ✗                                                                                                                                               80d ⚑  
> g++ const_string_vs_value_string.cpp -O3 -std=c++11

~/Desktop/cpp-benchs  master ✗                                                                                                                                               80d ⚑  
> ./a.out                                            
function_a: 0.002ms         # const ref 
function_b: 68.776ms       # copy
function_b-move: 2.86ms # move
```

Comment (blair1618) [inline core/src/scene/styleMixer.cpp]: Fair enough, refs it is.

**Comment by tallytalwar** [comment]
Updated.

**Comment by blair1618** [comment]
LGTM

**Comment by karimnaaji** [comment]
Looks good to me too, thanks for fixing this @tallytalwar

**Pull request merged**